src/NumericalAlgorithms/Integration/GslQuadAdaptive.hpp

```diff
@@ -0,0 +1,261 @@
+
+struct Result {
+  double value;
+  double error;
+};
+
+namespace detail {
+// forward decleration
+template <typename... Args>
+class GslQuadAdaptiveImpl;
+
+// ExpandTupleImpl takes and unpacks all args to pass them to the integrand,
+// the Algorithms in GSL Quad do not allow for a lambda function to be passed
+template <typename Args, typename ArgsIndices =
+                             std::make_index_sequence<tmpl::size<Args>::value>>
+struct ExpandTupleImpl;
+
+template <typename... Args, size_t... ArgsIndices>
+struct ExpandTupleImpl<tmpl::list<Args...>,
+                       std::index_sequence<ArgsIndices...>> {
+  template <typename FunctionType>
+  static double apply(FunctionType&& function, const double x,
+                      const std::tuple<Args...>& args) {
+    return (*function)(x, std::get<ArgsIndices>(args)...);
+  }
+};
+
+template <typename... Args>
+double integrand(double x, void* params) {
+  auto args = reinterpret_cast<GslQuadAdaptiveImpl<Args...>*>(params);
+  return ExpandTupleImpl<tmpl::list<Args...>>::apply(args->function_, x,
+                                                     args->args_);
+}
+
+template <typename... Args>
+class GslQuadAdaptiveImpl {
+ public:
+  GslQuadAdaptiveImpl(size_t max_intervals) noexcept
+      : max_intervals_(max_intervals) {
+    initialize();
+  }
+
+  GslQuadAdaptiveImpl() = default;
+  GslQuadAdaptiveImpl(const GslQuadAdaptiveImpl&) = default;
+  GslQuadAdaptiveImpl& operator=(const GslQuadAdaptiveImpl&) = default;
+  GslQuadAdaptiveImpl(GslQuadAdaptiveImpl&&) noexcept = default;
+  GslQuadAdaptiveImpl& operator=(GslQuadAdaptiveImpl&& rhs) noexcept = default;
+  ~GslQuadAdaptiveImpl() noexcept = default;
+
+  // clang-tidy: no runtime references
+  void pup(PUP::er& p) noexcept {
+    p | max_intervals_;
+    if (p.isUnpacking()) {
+      initialize();
+    }
+  }
+
+  double (*function_)(double, Args...);
+  std::tuple<Args...> args_;
+
+  template <size_t Index, typename Arg>
+  void set_parameter(const Arg& arg) {
+    std::get<Index>(args_) = arg;
+  }
+
+  template <typename IntegrandType>
+  void set_integrand(IntegrandType&& integrand) {
+    function_ = integrand;
+  }
+
+ private:
+  struct gsl_integration_workspace_deleter {
+    void operator()(gsl_integration_workspace* workspace) const noexcept {
+      gsl_integration_workspace_free(workspace);
+    }
+  };
+
+  void initialize() noexcept {
+    workspace_ = std::unique_ptr<gsl_integration_workspace,
+                                 gsl_integration_workspace_deleter>{
+        gsl_integration_workspace_alloc(max_intervals_)};
+    integrand_.function = &detail::integrand<Args...>;
+    integrand_.params = this;
+  }
+
+ protected:
+  size_t max_intervals_;
+  gsl_function integrand_;
+
+  std::unique_ptr<gsl_integration_workspace, gsl_integration_workspace_deleter>
+      workspace_;
+};
+}  // namespace detail
+
+/*!
+ * \ingroup NumericalAlgorithmsGroup
+ * \brief A adaptive integration procedure class
```

Comment:
Perhaps `A wrapper around the GSL adaptive integration procedures`

Please move this documentation above line 138 so it is correctly associated with the `GslQuadAdaptive` class